### PR TITLE
Bug 1830270: cmd: add explain subcommand

### PR DESCRIPTION
**Pull request by abhinavdahiya**
```console
$ ./bin/openshift-install explain installconfig
KIND:     InstallConfig
VERSION:  v1

RESOURCE: <object>
  InstallConfig is the configuration for an OpenShift install.

FIELDS:
    additionalTrustBundle <string>
      AdditionalTrustBundle is a PEM-encoded X.509 certificate bundle that will be added to the nodes' trusted certificate store.

    apiVersion <string>
      APIVersion defines the versioned schema of this representation of an object. Servers should convert recognized schemas to the latest internal value, and may reject unrecognized values. More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#resources

    baseDomain <string> -required-
      BaseDomain is the base domain to which the cluster should belong.

    compute <[]object>
      Compute is the configuration for the machines that comprise the compute nodes.
      MachinePool is a pool of machines to be installed.

    controlPlane <object>
      ControlPlane is the configuration for the machines that comprise the control plane.

    fips <boolean>
      Default: false
      FIPS configures https://www.nist.gov/itl/fips-general-information

    imageContentSources <[]object>
      ImageContentSources lists sources/repositories for the release-image content.
      ImageContentSource defines a list of sources/repositories that can be used to pull content.

    kind <string>
      Kind is a string value representing the REST resource this object represents. Servers may infer this from the endpoint the client submits requests to. Cannot be updated. In CamelCase. More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#types-kinds

    metadata <object> -required-
      <empty>

    networking <object>
      Networking is the configuration for the pod network provider in the cluster.

    platform <object> -required-
      Platform is the configuration for the specific platform upon which to perform the installation.

    proxy <object>
      Proxy defines the proxy settings for the cluster. If unset, the cluster will not be configured to use a proxy.

    publish <string>
      Default: "External"
      Valid Values: "","External","Internal"
      Publish controls how the user facing endpoints of the cluster like the Kubernetes API, OpenShift routes etc. are exposed. When no strategy is specified, the strategy is `External`.

    pullSecret <string> -required-
      PullSecret is the secret to use when pulling images.

    sshKey <string>
      SSHKey is the public Secure Shell (SSH) key to provide access to instances.
```
```console
$ ./bin/openshift-install explain installconfig.platform.aws
KIND:     InstallConfig
VERSION:  v1

RESOURCE: <object>
  AWS is the configuration used when installing on AWS.

FIELDS:
    amiID <string>
      AMIID is the AMI that should be used to boot machines for the cluster. If set, the AMI should belong to the same region as the cluster.

    defaultMachinePlatform <object>
      DefaultMachinePlatform is the default configuration used when installing on AWS for machine pools which do not define their own platform configuration.

    region <string> -required-
      Region specifies the AWS region where the cluster will be created.

    serviceEndpoints <[]object>
      ServiceEndpoints list contains custom endpoints which will override default service endpoint of AWS Services. There must be only one ServiceEndpoint for a service.
      ServiceEndpoint store the configuration for services to override existing defaults of AWS Services.

    subnets <[]string>
      Subnets specifies existing subnets (by ID) where cluster resources will be created.  Leave unset to have the installer create subnets in a new VPC on your behalf.

    userTags <object>
      UserTags additional keys and values that the installer will add as tags to all resources that it creates. Resources created by the cluster itself may not include these tags.
```
```console
$ ./bin/openshift-install explain installconfig.platform.libvirt
KIND:     InstallConfig
VERSION:  v1

RESOURCE: <object>
  Libvirt is the configuration used when installing on libvirt.

FIELDS:
    URI <string>
      Default: "qemu+tcp://192.168.122.1/system"
      URI is the identifier for the libvirtd connection.  It must be reachable from both the host (where the installer is run) and the cluster (where the cluster-API controller pod will be running). Default is qemu+tcp://192.168.122.1/system

    defaultMachinePlatform <object>
      DefaultMachinePlatform is the default configuration used when installing on libvirt for machine pools which do not define their own platform configuration. Default will set the image field to the latest RHCOS image.

    network <object>
      Network
```
```console
$ ./bin/openshift-install explain installconfig.platform.aws.serviceEndpoints
KIND:     InstallConfig
VERSION:  v1

RESOURCE: <[]object>
  ServiceEndpoints list contains custom endpoints which will override default service endpoint of AWS Services. There must be only one ServiceEndpoint for a service.

FIELDS:
    name <string>
      Name is the name of the AWS service. This must be provided and cannot be empty.

    url <string>
      URL is fully qualified URI with scheme https, that overrides the default generated endpoint for a client. This must be provided and cannot be empty.
```

### CODE REVIEW

**Comment by abhinavdahiya** [comment]
> f036749 
has a replace directive to local fork, because of https://github.com/kubernetes-sigs/controller-tools/pull/427

otherwise the `pkg.IPNet` prevents the creation of the CRD object...
/wip

File: docs/dev/explain.md [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+The generator uses the Godoc comments on the fields and types to create the descriptions. Therefore, making sure that everything used by the `InstallConfig` definition has user friendly comments will automatically transfer to user friendly explain output.
+
+## Kubebuilder markers
+
+The generator allows use of various markers for fields and types to provide information about the valid inputs. Various available markers are defined [here][kubebuilder-validation-markers]
+
+## Additional guidelines on godoc comments
+
+All definitions in installer codebase already follow and enforce [effective-go] guidelines, but for better explain output for the types these additional guidelines should also be followed
```
Comment (wking) [inline docs/dev/explain.md:26]: missing an anchor for `effective-go`?

File: docs/dev/explain.md [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+1. No external types are allowed to be used in the `InstallConfig`. All the types used should be defined in the installer repository itself. The only exception is the `TypeMeta` and `ObjectMeta`.
+
+2. Fields should always have `json` tags that follow [mixedCaps][go-mixed-caps] and should always start with lowercase.
+
+3. The comments on the fields must use the `json` tag to reference the field.
+
+4. Optional fields must have the `+optional` marker defined. Optional fields must also define the default value. If the values is static, `+kubebuilder:validation=Default` marker should be used, otherwise the comment must clearly define how the default value will be calculated.
```
Comment (wking) [inline docs/dev/explain.md:34]: nit: "values is" -> "value is" or "values are"

**Comment by sdodson** [comment]
/retitle Bug 1830270: cmd: add explain subcommand

File: pkg/explain/cmd.go [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+// NewCmd returns a subcommand for explain
+func NewCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "explain",
+		Short: "List the fields for supported InstallConfig versions",
+		Long: `This command describes the fields associated with each supported InstallConfig API. Fields are identified via a simple
+JSONPath identifier:
+		
+isntallconfig.<fieldName>[.<fieldName>]
```
Comment (patrickdillon) [inline pkg/explain/cmd.go:21]: `isntallconfig` -> `installconfig`

File: pkg/explain/cmd.go [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+	file, err := data.Assets.Open(installConfigCRDFileName)
+	if err != nil {
+		return errors.Wrap(err, "failed to load InstallConfig CRD")
+	}
+	defer file.Close()
+
+	raw, err := ioutil.ReadAll(file)
+	if err != nil {
+		return errors.Wrap(err, "faield to read InstallConfig CRD")
```
Comment (patrickdillon) [inline pkg/explain/cmd.go:51]: `faield` -> `failed`

File: docs/dev/explain.md (hunk 2 above)
Comment (abhinavdahiya) [reply docs/dev/explain.md]: @wking fixed

File: docs/dev/explain.md (hunk 1 above)
Comment (abhinavdahiya) [reply docs/dev/explain.md:26]: @wking fixed

File: pkg/explain/cmd.go (hunk 3 above)
Comment (abhinavdahiya) [reply pkg/explain/cmd.go]: @patrickdillon fixed

File: pkg/explain/cmd.go (hunk 4 above)
Comment (abhinavdahiya) [reply pkg/explain/cmd.go]: @patrickdillon fixed

**Comment by patrickdillon** [comment]
LGTM 
Needs rebase but there is no label. Ping me for lgtm when resolved.

**Comment by abhinavdahiya** [comment]
@patrickdillon rebased the PR on master

**Pull request merged**